Add ability to automagically determine which model config to load

d8ahazard opened the pull request:
Might be a bit heavy for lower-VRAM users, and it also depends on the global step of the source model not changing, but it does work with both the base 512 and 768 models, plus non v2 models. No YAML files required. 

Also, might be a good idea to put the configs in their own subdirectory, but I was trying to keep this short and sweet.

d8ahazard committed:
Auto-load model config.

d8ahazard committed:
That should do it.

leppie commented on modules/sd_models.py:
```diff
@@ -227,20 +229,43 @@ def load_model_weights(model, checkpoint_info, vae_file="auto"):
+    try:
+        checkpoint = torch.load(path)
+        c_dict = checkpoint["state_dict"]
+        v2_key = "model.diffusion_model.input_blocks.2.1.transformer_blocks.0.attn2.to_k.weight"
```
Looks like this key is present in all models. 

This is only in V2 models from what I can see:

```
v2_key = "cond_stage_model.model.ln_final.weight"
```

d8ahazard replied:
Good call. Used your key - I'll also add this to my conversion scripts for dreambooth, as that's where this came from.

Also added a bit to ensure we don't load the checkpoint twice if calling from reload_model_weights, a check in the whole checkpoint for the key if the model was created without a "state_dict" entry, and made the log message a bit more useful.

Tested with the v2.0 768-v checkpoint, v2.0 512 checkpoint, several v1x checkpoints, and the new v2.1 checkpoints. All load as intended with no yaml files needed.

leppie replied:
How do you differenciate between the 512 and 768 models with the updated code? I dont see anything special dealing with that.

d8ahazard replied:
It's the number of training steps. It's really the only difference between them, and it's the hack Huggingface suggested to use...

Need to find the number of steps for the 512 v2.1 model - but only if the 2.1 512 doesn't use v-prediction.

d8ahazard committed:
Use better V2 Key, print cleanup, support models without "state_dict" key.

Despite our conversion script relying on the k.weight key, use ln_final.weight instead to determine if model is V2.
Print a more intelligent and useful message.
Fix loading checkpoint twice if called from reload model weights.
Ensure models that don't have state dict under the "state_dict" key work.

d8ahazard committed:
v2.1 support

Need to not hijack model if 2.1, as well as bump SD repo.

leppie commented:
Not sure what the cause is, but the last commit breaks image upscaling in img2img.

Repro:

1. Use v1.5 model, gen 512x512 image in txt2img
2. Send to img2img
3. Change size to 1024x1024
4. Generate
5. KabOOM (pun intended)

```
  File "G:\stable-diffusion-webui\repositories\stable-diffusion-stability-ai\ldm\modules\attention.py", line 177, in forward
    sim = einsum('b i d, b j d -> b i j', q, k) * self.scale
  File "g:\stable-diffusion-webui\venv\lib\site-packages\torch\functional.py", line 360, in einsum
    return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
RuntimeError: CUDA out of memory. Tried to allocate 16.00 GiB (GPU 0; 10.00 GiB total capacity; 2.22 GiB already allocated; 5.23 GiB free; 2.30 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

Not sure why it is asking for 16GB! Without [2f25c20](https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/5493/commits/2f25c20bf6711cfd2d5499a0cdd417401a5eeca2), it works just fine. Only uses about 3GB.

Args: `--ckptfix --xformers`

d8ahazard committed:
Revert hijack skip, bump diffusers, add ver check for v2.1 512x model.

With diffusers 0.10.2, we don't need xformers enabled or to skip hijacking stuff in order to use v2.1 models.

Add version check to load proper config for v2.1 512x model.

d8ahazard committed:
Remove unnecessary V21 checks.

We don't care with diffusers 0.10.2. They fix it for us.

AUTOMATIC1111 commented:
> if checkpoint["global_step"] == 875000 or checkpoint["global_step"] == 220000:

I rather not have this. Just let the user put a config next to the model.
